<a href="https://colab.research.google.com/github/Soot3/Spotify-Analysis/blob/main/Spotify_NG(scrape).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
!pip install spotipy
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials

In [3]:
authorization={"client_id": "xxxx", "client_secret": "xxxx"}

In [4]:
client_credentials_manager = SpotifyClientCredentials(client_id=authorization['client_id'],client_secret=authorization['client_secret'])

sp = spotipy.Spotify(client_credentials_manager=client_credentials_manager)

Spotify is not available in Nigeria, however, Nigerian music has a place on the streaming platform as Nigerian musicians post their music taking advantage of its reach.

Using a few spotify playlists which are based on Nigerian musicians and Nigerian music, we can gain access and analyse Nigerian music as a whole.

In [5]:
# Function gets data on the Spotify track and artist ids from the given playlist id
def getTrackIDs(playlist_id):
    track_ids = []
    artist_ids = []
    playlist = sp.playlist(playlist_id)
    for item in playlist['tracks']['items']:
        track = item['track']
        artist = track['artists'][0]['external_urls']['spotify']
        track_ids.append(track['id'])
        artist_ids.append(artist)
    return track_ids, artist_ids

In [7]:
#Looping through our chosen playlists
links = ['https://open.spotify.com/playlist/0y6GLcb6chq3IXB63zImHA?si=LVVhdiiLQqCjbkZKrIrRqQ','https://open.spotify.com/playlist/0yTBSYuj6a4vQScMLMZG6V?si=mXutCuNjRNuQzhAVzMoR3g','https://open.spotify.com/playlist/1PWKPQG5CWNKlMTbtCn8Xf?si=gAJvNC1PSByXpkgGNYC2Fg','https://open.spotify.com/playlist/4mRsLwJcOSpAPhFj1lT5uH?si=04botuN9S_-3FX1C2uZf0A','https://open.spotify.com/playlist/5pJRsC8aY41B2FuWnrSJBN?si=ty14WQw0Qe-a83CNJrtvcg','https://open.spotify.com/playlist/4TnTkspo7B9C03rjOoAoLW?si=8BJrK0GJRtO49JdcIm8Pzw','https://open.spotify.com/playlist/23fCAKOF4pG6E0baakihsK?si=5B292X27RYaPGPp4oibBVw']
mt_list = []
artist_list = []
for i in links:
  ids, artist = getTrackIDs(i)
  mt_list.append(ids)
  artist_list.append(artist)

In [8]:
# The funtion collates indepth song data
def getTrackFeatures(id,a):
  meta = sp.track(id)
  features = sp.audio_features(id)
  artist_data = sp.artist(a)

  # meta
  name = meta['name']
  album = meta['album']['name']
  artist = meta['album']['artists'][0]['name']
  if artist_data['genres'] == []:
    genre = 'Missing'
  else:
    genre = artist_data['genres'][0]
  release_date = meta['album']['release_date']
  length = meta['duration_ms']
  popularity = meta['popularity']

  # features
  acousticness = features[0]['acousticness']
  danceability = features[0]['danceability']
  energy = features[0]['energy']
  instrumentalness = features[0]['instrumentalness']
  liveness = features[0]['liveness']
  loudness = features[0]['loudness']
  speechiness = features[0]['speechiness']
  tempo = features[0]['tempo']
  time_signature = features[0]['time_signature']

  track = [name, album, artist, genre, release_date, length, popularity, danceability, acousticness, energy, instrumentalness, liveness, loudness, speechiness, tempo, time_signature]
  return track

In [11]:
# Flatten the data from a list of lists structure
flat_list = [item for sublist in mt_list for item in sublist]
flat_list1 = [item for sublist in artist_list for item in sublist]

In [12]:
# loop over track ids 
tracks = []
for j in range(len(flat_list1)):
  track = getTrackFeatures(flat_list[j], flat_list1[j])
  tracks.append(track)

# create dataset
df = pd.DataFrame(tracks, columns = ['name', 'album', 'artist', 'artist_top_genre', 'release_date', 'length', 'popularity', 'danceability', 'acousticness', 'energy', 'instrumentalness', 'liveness', 'loudness', 'speechiness', 'tempo', 'time_signature'])

In [13]:
df.head()

name  ... time_signature
0                    Sparky  ...              5
1                shuga rush  ...              3
2                     LITT!  ...              4
3  Confident / Feeling Cool  ...              4
4                wanted you  ...              4

[5 rows x 16 columns]

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 576 entries, 0 to 575
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   name              576 non-null    object 
 1   album             576 non-null    object 
 2   artist            576 non-null    object 
 3   artist_top_genre  576 non-null    object 
 4   release_date      576 non-null    object 
 5   length            576 non-null    int64  
 6   popularity        576 non-null    int64  
 7   danceability      576 non-null    float64
 8   acousticness      576 non-null    float64
 9   energy            576 non-null    float64
 10  instrumentalness  576 non-null    float64
 11  liveness          576 non-null    float64
 12  loudness          576 non-null    float64
 13  speechiness       576 non-null    float64
 14  tempo             576 non-null    float64
 15  time_signature    576 non-null    int64  
dtypes: float64(8), int64(3), object(5)
memory us

In [16]:
df.drop_duplicates(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 530 entries, 0 to 575
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   name              530 non-null    object 
 1   album             530 non-null    object 
 2   artist            530 non-null    object 
 3   artist_top_genre  530 non-null    object 
 4   release_date      530 non-null    object 
 5   length            530 non-null    int64  
 6   popularity        530 non-null    int64  
 7   danceability      530 non-null    float64
 8   acousticness      530 non-null    float64
 9   energy            530 non-null    float64
 10  instrumentalness  530 non-null    float64
 11  liveness          530 non-null    float64
 12  loudness          530 non-null    float64
 13  speechiness       530 non-null    float64
 14  tempo             530 non-null    float64
 15  time_signature    530 non-null    int64  
dtypes: float64(8), int64(3), object(5)
memory us

Our dataset has been collected with 530 unique entries

In [17]:
df['release_date'].nunique()

393

The formatting of release dates does not need to include the specific day and month, as it needlessly inflates the release date variable

In [18]:
df['release_date'] = pd.to_datetime(df['release_date'])
df['release_date'] = pd.DatetimeIndex(df['release_date']).year
df['release_date'].nunique()

20

We have at least one song for two decades in this dataset

In [19]:
df.head()

name  ... time_signature
0                    Sparky  ...              5
1                shuga rush  ...              3
2                     LITT!  ...              4
3  Confident / Feeling Cool  ...              4
4                wanted you  ...              4

[5 rows x 16 columns]

In [20]:
df['release_date'].value_counts()

2016    95
2017    83
2014    75
2015    55
2019    51
2018    47
2020    25
2013    25
2012    22
2011    16
2009    11
2010     9
2008     7
2004     2
2003     2
2007     1
2006     1
2001     1
2000     1
1998     1
Name: release_date, dtype: int64

The data can be limited to 2010 downwards as the rest occupy very little space in the dataset

In [ ]:
df['artist_top_genre'].unique()

array(['alternative r&b', 'afropop', 'indie r&b', 'nigerian pop',
       'afro r&b', 'afro dancehall', 'Missing', 'azonto', 'soft rock',
       'hiplife', 'erotica', 'nigerian hip hop', 'highlife',
       'azontobeats', 'south african house', 'bongo flava', 'afroswing',
       'gqom', 'christian afrobeat'], dtype=object)

In [21]:
df['artist_top_genre'].value_counts()

afro dancehall         328
afropop                 90
Missing                 53
nigerian pop            24
azontobeats              8
alternative r&b          7
indie r&b                4
nigerian hip hop         3
afro r&b                 3
soft rock                1
highlife                 1
gqom                     1
south african house      1
erotica                  1
bongo flava              1
azonto                   1
christian afrobeat       1
hiplife                  1
afroswing                1
Name: artist_top_genre, dtype: int64

Again we see a few categories having a large percentage of the entries in the dataset. Also, 53 of the entries in the entire dataset have no genre data attached, marking them as Missing (the third highest category)

In [ ]:
df[df['artist_top_genre'] == 'bongo flava']

name       album  ...    tempo time_signature
401  Baby Oku (Gyration)  Super Sexy  ...  104.178              3

[1 rows x 16 columns]

Checking a few esoteric genres 

In [ ]:
df.to_csv('nigerian_spotify_songs1.csv', index=False, encoding='utf-8')